pkg/services/publicdashboards/service/query.go

```diff
@@ -319,3 +322,54 @@ func sanitizeMetadataFromQueryData(res *backend.QueryDataResponse) {
 		}
 	}
 }
+
+// NewDataTimeRange declared to be able to stub this function in tests
+var NewDataTimeRange = legacydata.NewDataTimeRange
+
+// BuildTimeSettings build time settings object using selected values if enabled and are valid or dashboard default values
+func buildTimeSettings(d *dashboards.Dashboard, reqDTO models.PublicDashboardQueryDTO, pd *models.PublicDashboard) models.TimeSettings {
+	from := d.Data.GetPath("time", "from").MustString()
+	to := d.Data.GetPath("time", "to").MustString()
+	tz := d.Data.GetPath("timezone").MustString()
+
+	location := getLocation(reqDTO, tz)
+	if pd.TimeSelectionEnabled && reqDTO.TimeRange.From != "" && reqDTO.TimeRange.To != "" {
+		from = reqDTO.TimeRange.From
+		to = reqDTO.TimeRange.To
+	}
+
+	timeRange := NewDataTimeRange(from, to)
+
+	timeFrom, _ := timeRange.ParseFrom(
+		legacydata.WithLocation(location),
+	)
+	timeTo, _ := timeRange.ParseTo(
+		legacydata.WithLocation(location),
+	)
+	timeToAsEpoch := timeTo.UnixMilli()
+	timeFromAsEpoch := timeFrom.UnixMilli()
+
+	// Were using epoch ms because this is used to build a MetricRequest, which is used by query caching, which want the time range in epoch milliseconds.
+	return models.TimeSettings{
+		From: strconv.FormatInt(timeFromAsEpoch, 10),
+		To:   strconv.FormatInt(timeToAsEpoch, 10),
+	}
+}
+
+// getLocation returns the location from the request or the default location
+func getLocation(reqDTO models.PublicDashboardQueryDTO, dashboardLocation string) *time.Location {
+	// if the Location is blank or there is an error default is UTC
+	location, err := time.LoadLocation(dashboardLocation)
+	if err != nil {
+		location = time.UTC
+	}
+
+	if reqDTO.TimeRange.Location != "" {
+		userLocation, err := time.LoadLocation(reqDTO.TimeRange.Location)
+		if err == nil {
+			location = userLocation
+		}
+	}
+
+	return location
```

Comment:
Instead of unnecessary calculating the location with time.LoadLocation if the param Location is not null , I would do something like this

```suggestion
	if reqDTO.TimeRange.Location != "" {
		if userLocation, err := time.LoadLocation(reqDTO.TimeRange.Location); err == nil {
			return userLocation
		}
	}
	
	// if the Location is blank or there is an error default is UTC
	if location, err := time.LoadLocation(dashboardLocation); err != nil {
        return time.UTC
	}
	
	return location
```